File: 973-k-closest-points-to-origin/973-k-closest-points-to-origin.java
class Solution {
    public int[][] kClosest(int[][] points, int k) {
        PriorityQueue<int[]> pq = new PriorityQueue<>((a, b) -> - (a[0] * a[0] + a[1] * a[1]) + (b[0] * b[0] + b[1] * b[1]));
        
        for(int i = 0; i < points.length; i++) {
            int[] p = points[i];
            pq.offer(p);
            
            while(pq.size() > k) {
                pq.poll();
            }
        }
        
        int[][] result = new int[k][2];
        
        for(int i = 0; i < k; i++) {
            result[i] = pq.poll();
        }
        
        return result;
    }
}

File: 973-k-closest-points-to-origin/973-k-closest-points-to-origin.scala
import scala.collection.mutable._
object Solution {
    def kClosest(points: Array[Array[Int]], k: Int): Array[Array[Int]] = {
        
        val pq: PriorityQueue[Array[Int]] = PriorityQueue()(Ordering.by { p => p(0) * p(0) + p(1) * p(1) })

        for(p <- points) {
            pq.enqueue(p)
            while(pq.length > k) pq.dequeue
        }

        var res: Array[Array[Int]] = Array()

        while(pq.nonEmpty) {
            res = res :+ pq.dequeue
        }

        res
    }
}